### PR TITLE
Enable formatting for new syntax changes and fix formatting issues

### LINKED ISSUE #25786: Add formatting support for 'on fail' clause
**Description:**
Refer: https://ballerina.io/ballerina-spec/spec.html#on-fail-clause

### LINKED ISSUE #25788: Add formatter support for 'isolate' changes

### LINKED ISSUE #25865: [Formatter] Support the new object syntax changes
**Description:**
Support the object syntax changes (module class definition and object constructor expression) in the formatter.

* Object Type Def

**Previous:**
```ballerina
type Obj object {    
    int i;    
    function init(int i) {        
        self.i = i;    
    }
};
```
**Current:**
```ballerina
class Obj {   
    int i;    
    function init(int i) {        
        self.i = i;    
    }
}
```

* Abstract Objects

**Previous:**
```ballerina
type Obj abstract object {    
    function foo(int i);
};
```
**Current:**
```ballerina
type Obj object {    
    function foo(int i);
};
```

* Anon Objects

**Previous:**
```ballerina
function bar() {    
    object { 
        function foo() { 
            // do something 
        } 
    } o = new();
}
```
**Current:**
```ballerina
function bar() {    
    var o = object { 
        function foo() { 
            // do something 
        }
    };
}
```
**Steps to reproduce:**

**Affected Versions:**

**OS, DB, other environment details and versions:**

**Related Issues (optional):**
https://github.com/ballerina-platform/ballerina-lang/issues/25643

**Suggested Labels (optional):**

**Suggested Assignees (optional):**

### PR DESCRIPTION
**Pull request by IrushiL**
## Purpose

1. Enables formatting support for the following.
   * Object Constructor Expressions (abstract objects and anon objects)
   * Do Statement
   * On Fail Clause:
      * Do Statements
      * Match Statements
      * Foreach Statements
      * While Statements
      * Lock Statements
   * Isolated Functions

2. Fix some formatting issues.
3. Handle comments in the source code.

Fixes #25865
Fixes #25788
Fixes #25786

## Check List 
- [X] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

**Review by nadeeshaan — APPROVED** [review]
LGTM

File: misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/FormatterUtils.java [hunk 1]
```diff
@@ -244,15 +290,90 @@ static Token formatToken(Token token, int leadingSpaces, int trailingSpaces, int
...
+            if (parent.children().size() > childLocation + 1) {
+                Token nextToken = getFirstToken(parent.children().get(childLocation + 1));
+                if (nextToken != null && nextToken.containsLeadingMinutiae()) {
+                    MinutiaeList minutiaes = nextToken.leadingMinutiae();
+                    if (commentCount(minutiaes) > 0) {
+                        for (Minutiae minutiae : minutiaes) {
+                            if (minutiae.kind() == SyntaxKind.END_OF_LINE_MINUTIAE) {
+                                count++;
+                            } else {
+                                break;
```
Comment (SupunS) [inline misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/FormatterUtils.java:349]: Can invert the conditions and do an early exit.

File: misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/FormatterUtils.java [hunk 2]
```diff
@@ -410,6 +535,18 @@ private static Indentation getBlockParent(Node node) {
         return new Indentation(null, false);
     }
 
+    static boolean addNewTrailingLine(NonTerminalNode parent, NonTerminalNode node) {
+        if (parent != null) {
+            int childLocation = getChildLocation(parent, node);
+            if (parent.children().size() > childLocation + 1) {
+                Token nextToken = getFirstToken(parent.children().get(childLocation + 1));
+                if (nextToken != null && nextToken.containsLeadingMinutiae()) {
+                    return (nextToken.leadingMinutiae().get(0).kind() != SyntaxKind.END_OF_LINE_MINUTIAE);
```
Comment (SupunS) [inline misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/FormatterUtils.java:544]: Same as above

File: misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/FormatterUtils.java [hunk 3]
```diff
@@ -134,6 +167,14 @@ static DiagnosticPos getPosition(Node node) {
                 return parent;
             }
         }
+        if (parentKind == (SyntaxKind.OBJECT_CONSTRUCTOR) && grandParent != null &&
+                grandParent.kind() == (SyntaxKind.LOCAL_VAR_DECL)) {
```
Comment (SupunS) [inline misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/FormatterUtils.java:171]: Additional parenthesis around the syntax kind

File: misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/FormatterUtils.java [hunk 4]
```diff
@@ -134,6 +167,14 @@ static DiagnosticPos getPosition(Node node) {
...
+            return grandParent;
+        }
+        if (parentKind == (SyntaxKind.TYPE_CAST_PARAM) && grandParent != null &&
+                grandParent.kind() == (SyntaxKind.TYPE_CAST_EXPRESSION)) {
```
Comment (SupunS) [inline misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/FormatterUtils.java:175]: same as above

File: misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/FormatterUtils.java [hunk 5]
```diff
@@ -244,15 +290,90 @@ static Token formatToken(Token token, int leadingSpaces, int trailingSpaces, int
...
+                }
+            }
+        }
+        return minutiaes;
+    }
+
+    private static int commentCount(MinutiaeList minutiaeList) {
+        int count = 0;
+        for (int i = 0; i < minutiaeList.size(); i++) {
+            if (minutiaeList.get(i).kind() == SyntaxKind.COMMENT_MINUTIAE) {
```
Comment (nadeeshaan) [inline misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/FormatterUtils.java:329]: Shall we invert the condition with a continue?

Comment (IrushiL) [reply misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/FormatterUtils.java:329]: This method checks the number of comment minutiaes available inside the minutiae list. Therefore, the entire list should be looped.

File: misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/FormatterUtils.java (hunk 1 above)
Comment (IrushiL) [reply misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/FormatterUtils.java]: Done

File: misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/FormatterUtils.java (hunk 2 above)
Comment (IrushiL) [reply misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/FormatterUtils.java]: Done

File: misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/FormatterUtils.java (hunk 3 above)
Comment (IrushiL) [reply misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/FormatterUtils.java]: Fixed

File: misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/FormatterUtils.java (hunk 4 above)
Comment (IrushiL) [reply misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/FormatterUtils.java]: Fixed

**Review by SupunS — APPROVED** [review]

**Pull request merged**